### PR TITLE
fixing FindRoot

### PR DESCRIPTION
**Pull request by mmatera**
This seems to work.

TODO
---------

* Compile Functions and Derivatives
* Implementing other methods
* Implementing checks in AccuracyGoal  and PrecisionGoal
* FindRoot over vectors (F is a vector, or x0 is a vector)

### CODE REVIEW

File: mathics/builtin/calculus.py [hunk 1]
```diff
@@ -1039,6 +1040,68 @@ def apply(self, f, n, n0, evaluation, options={}):
         except:
             pass
 
+def find_root_secant(f, x0, x, opts, evaluation) -> (Number, bool):
```
Comment (rocky) [inline mathics/builtin/calculus.py:1043]: Is it not possbile under various conditions to hook into mpmath's routines for doing this? See https://mpmath.org/doc/current/calculus/optimization.html for example.

If we do so we would I suspect we would improve performance over, immediately provide more solvers, while reducing the chance of there being an error that we introduce.

Comment (rocky) [reply mathics/builtin/calculus.py:1043]: And in the improtant case where the end is Equal to 0 or some constant, can't we use Sympy's solve. See https://docs.sympy.org/latest/modules/solvers/solvers.html

Comment (mmatera) [reply mathics/builtin/calculus.py:1043]: Indeed, the idea is to do the same as in NIntegrate. But I wanted to have a couple of "native" routines when the optimized routines are not available.

Comment (rocky) [reply mathics/builtin/calculus.py:1043]: Ok, that makes sense. But it doesn't look like we are trying to use any of the sympy or mathmp routines that core these. Is this correct?

Comment (rocky) [reply mathics/builtin/calculus.py:1043]: I haven't looked at the corresponding mpmath or sympy math routines. How do they compare with what is here code-wise?

Comment (mmatera) [reply mathics/builtin/calculus.py:1043]: Before delegating the work on that routines, we should check if it is possible to pre-compile the target function, and its derivatives,  in a way that we can pass those compiled versions. Also, for the one-dimensional problem, the implemented algorithms are quite efficient, having a bottleneck in the function evaluation. On the other hand, FindRoot with vector variables is not implemented yet, so I think that the use of mpmath routines should be implemented at a later stage.

**Review by rocky — APPROVED** [review]
LGTM 

Personally I think hooking into sympy would be really nice. And would be nice to benchmark. 

But I will add this to an issue.

**Comment by rocky** [comment]
@mmatera many thanks for fixing.

**Pull request merged**